### DISCUSSION TITLE
Can the coverage map of the Sionna RT record planes of multiple heights at the same time？

### DISCUSSION THREAD

**Discussion by yzw0905** [NVlabs/sionna#230] [Q&A] [answered]
Hi, I am a student who is learning sionna RT, I would like to ask if it is possible to generate a coverage map in space, similar to recording the gain of a plane at multiple heights at a base station in one simulation, can someone help me?

**Comment by jhoydis** [comment, marked as answer]
Hi,

In short, the answer is no. You can compute as many different coverage maps as you want but not in the same simulation run.
I am not sure if it helps, but you can define coverage maps with arbitrary rotation, e.g., vertical. Alternatively, you could place a large number of receivers at different heights and compute an approximate coverage map from the channel impulse responses..

**Reply by yzw0905** [reply]
Thank you for your reply, it helped me a lot.

**Reply by achaitu** [reply]
Hi, by "compute an approximate coverage map from the channel impulse responses", how is the coverage map strength computed from the channel impulse responses? Is it straight forward? Is there any source I could look up to in the documentation?

**Reply by jhoydis** [reply]
Yes, this is rather simple. The definition of coverage map we use can be found [here](https://nvlabs.github.io/sionna/api/rt.html#coverage-map). In principle, the coverage map tries to average out small scale fading over small geographic regions.

Assume that you would place one receiver in the centre of each tile of your coverage map. Then, the value of the coverage map can be approximated by the sum of the squared amplitudes of the path coefficients $a_i$. In code that would look like this:

```
a, tau = paths.cir()
pl = tf.reduce_sum(tf.abs(a)**2, axis=-2) # Summation over all paths
```

You could also place multiple receivers in each tile and average their individual path losses. This would give a better approximation of the coverage map, but it is also far more computationally intensive.

As a side remark, when you use the [``scene.coverage_map``](https://nvlabs.github.io/sionna/api/rt.html#coverage-map) function, a more efficient approach is used which does not explicitly simulate channel impulse responses at precise positions but rather integrates path contributions over the surface tiles of the coverage map.

**Reply by yzw0905** [reply]
Hi, before I asked how to accelerate rt on multiple GPUs, you gave a negative answer, but you mentioned that you can run multiple transmitters and receivers in parallel on multiple GPUs, I want to ask whether parallel means setting up their respective GPUs in multiple programs and then computing at the same time? Can you give some sample code

**Reply by jhoydis** [reply]
First of all, you can compute the channel impulse response for many transmitters and receivers in parallel on a single GPU. The main limiting factor is memory. 

If you have two GPUs available, you could run two different scripts or notebooks which in which you set the active GPU to a specific value. This could look like this:

```
import tensorflow as tf
gpus = tf.config.list_physical_devices('GPU')
print('Number of GPUs available :', len(gpus))
if gpus:
    gpu_num = 0 # Number of the GPU to be used
    try:
        tf.config.set_visible_devices(gpus[gpu_num], 'GPU')
        print('Only GPU number', gpu_num, 'used.')
        tf.config.experimental.set_memory_growth(gpus[gpu_num], True)
    except RuntimeError as e:
        print(e)
```

In one script, you would set ``gpu_num=0``, in the other ``gpu_num=1``.
